antlr_ast/ast.py

```diff
@@ -283,3 +176,505 @@ def reportContextSensitivity(
     ):
         return
         # raise Exception("TODO")
+
+
+# Parse Tree Visitor ----------------------------------------------------------
+# TODO: visitor inheritance not really needed, but indicates compatibility
+# TODO: make general node (Terminal) accessible in class property (.subclasses)?
+
+
+class BaseNode(AST):
+    """AST is subclassed so we can use Python ast module  visiting and walking on the custom AST"""
+
+    def __init__(self, children, field_references, label_references, ctx=None):
+        self.children = children
+
+        self._field_references = field_references
+        self.children_by_field = materialize(self._field_references, self.children)
+
+        self._label_references = label_references
+        self.children_by_label = materialize(self._label_references, self.children)
+
+        self._ctx = ctx
+
+    subclasses = {}
+
+    _fields = ()
+
+    # whether to descend for selection (greater descends into lower)
+    _priority = 2
+
+    # getattr: return None or raise for nonexistent attr
+    # in Transformer conditionals:
+    # - getattr(obj, attr, None) works with both
+    # - hasattr(obj, attr) if strict
+    # - obj.attr if not strict
+    _strict = False
+
+    @classmethod
+    def create(cls, ctx, children=None):
+        if children is None:
+            children = ctx.children
+
+        field_names = get_field_names(ctx)
+        children_by_field = get_field_references(ctx, field_names)
+
+        label_names = get_label_names(ctx)
+        children_by_label = get_field_references(ctx, label_names)
+
+        cls_name = type(ctx).__name__.split("Context")[0]
+        subclass = cls.get_node_cls(cls_name, tuple(field_names))
+        return subclass(children, children_by_field, children_by_label, ctx)
+
+    @classmethod
+    def get_node_cls(cls, cls_name, field_names=tuple()):
+        if cls_name not in cls.subclasses:
+            cls.subclasses[cls_name] = type(
+                cls_name, (BaseNode,), {"_fields": field_names}
+            )
+        return cls.subclasses[cls_name]
+
+    def __getattr__(self, name):
+        fallback = None
+        if self._strict:
+            fallback = AttributeError
```

Comment:
Setting `_strict` to `False` makes the code in the Transformer methods (the successor to 
AstVisitor methods) simpler: you can write `if obj.attr:` instead of `if hasattr(obj, attr)` or `if getattr(obj, "attr", None):`.
[This comment tries to explain the consequence of the `_strict` value](https://github.com/datacamp/antlr-ast/pull/15/files#diff-178691b73b690b4b29cf25e6d00d1902R207).

But because it can absorb errors, it's best to limit `_strict` being `False`.
That's why I set it to `True` on `AliasNode`, so as long as you are using the 'charted grammar territory' for which AliasNodes exist, attributes work as they work by default.

Reply:
I need to be efficient with my time, so want to be frank in my response: based on your comment, I don't know whether the answer to my question is yes or no.

Reply:
Ah, wait--I see, is it that this function won't return AttributeError? This seems like classic try/except territory, with the except doing something different depending on self._strict

Reply:
To your first question:
Yes, if it returns an AttributeError, the transformer methods will no longer work,
but it won't because on BaseNodes, `_strict` is false.

Reply:
The `children_by_label ... children_by_field` part should still be written as a nested `.get` with a fallback, but it could be in a try with the `None` vs `AttributeError` decision in the catch.